pandas/tests/reshape/test_concat.py

```diff
@@ -1771,6 +1772,148 @@ def test_concat_tz_series_tzlocal(self):
         tm.assert_series_equal(result, pd.Series(x + y))
         assert result.dtype == 'datetime64[ns, tzlocal()]'
 
+    def test_concat_NaT_dataframes_all_NaT_axis_0(self):
+        # GH 12396
+        expected = pd.DataFrame([pd.NaT, pd.NaT, pd.NaT], index=[0, 1, 0])
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.NaT]])
+
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
+        dtype = DatetimeTZDtype('ns', tz='UTC')
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]], dtype=dtype)
+
+        result = pd.concat([first, second], axis=0)
+        # upcasts for mixed case
+        expected = expected.apply(lambda x: x.astype(dtype))
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # both sides timezone-aware
+        second = pd.DataFrame([[pd.NaT]], dtype=dtype)
+
+        # upcasts to tz-aware
+        result = pd.concat([first, second], axis=0)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+    def test_concat_NaT_dataframes_all_NaT_axis_1(self):
+        # GH 12396
+        expected = pd.DataFrame([[pd.NaT, pd.NaT], [pd.NaT, pd.NaT]],
+                                columns=[0, 1])
+
+        # non-timezone aware
+        first = pd.DataFrame([[pd.NaT], [pd.NaT]])
+        second = pd.DataFrame([[pd.NaT]], columns=[1])
+
+        result = pd.concat([first, second], axis=1)
+        assert_frame_equal(result, expected, check_datetimelike_compat=True)
+
+        # one side timezone-aware
```

Comment:
i don't like using DatetimeTZDtype explicity like this its not user facing, instead

```
In [10]: pd.DataFrame(pd.Series([pd.NaT, pd.NaT]).dt.tz_localize('UTC'))
Out[10]: 
    0
0 NaT
1 NaT

In [11]: pd.DataFrame(pd.Series([pd.NaT, pd.NaT]).dt.tz_localize('UTC')).dtypes
Out[11]: 
0    datetime64[ns, UTC]
dtype: object
```